docs/rfc/identity_proofs.md

```diff
@@ -0,0 +1,232 @@
+to the proving side.
+
+We conclude that this mechanism would be a good fit for `radicle-link` (due to
+the similarities), and a desirable feature of applications built on top of it.
+
+The following sections describe how such claims shall be stored in the identity
+payload of a `radicle-link` identity, how to obtain a publishable proof, and how
+to verify it.
+
+## Claims
+
+Claims can only be made by identities of kind `Person`, and claim a single
+external account identifier. They are introduced by defining a new payload type,
+identified by the URL:
+
+    https://radicle.xyz/link/claim/v1
+
+The shape of the JSON structure is:
+
+```json
+{
+    "SERVICE": {
+        "account": "STRING"
+        "expiration": {
+            "created": INTEGER,
+            "expires": INTEGER
+        },
+        "proof": "URL"
+    }
+}
+```
+
+Where the fields denote:
+
+* `SERVICE`
+
+  A conventional identifier of the external service, e.g. "github", "twitter",
+  "radicle-ethereum"
+
+* `account`
+
+  The unique account identifier within the service, using the service-specific
+  canonical string representation e.g. "kim", "0x32be343b94f860124dc4fee278fdcbd38c102d88".
+
+* `expiration` (optional)
+    * `created`
+
+      Creation timestamp of the claim, in seconds since 1970-01-01T00:00:00Z.
+
+    * `expires`
+
+      Seconds relative to `created`, after which the claim should no longer be
+      considered.
+
+* `proof` (optional)
+
+  A URL to assist verification tooling in retrieving the proof from the external
+  system. This is mainly a convenience, and obviously requires creation of a new
+  revision after the fact.
+
+## Proof Generation
+
+The above claim payload is committed to the identity history as a new revision.
+Technically, this revision needs to be approved by all key delegations for
+verification to pass later on, but since we assume that eligible keys are held
+by the same person, it may be acceptable to publish the proof right away for
+user experience reasons.
+
+The actual proof consists of the following tuple:
+
+    (root, revision, public-key, signature)
+
+Note that the "git" protocol specifier of `radicle-link` URNs is implied, that
+is, future version MUST treat the absence of a disambiguating value as denoting
+"git".
+
+The values `root`, `revision`, and `public-key` are specified in
+[identities][ids], and it is RECOMMENDED to follow the serialisation formats
+devised there. `signature` is the Ed25519 signature over `revision`, in much the
+same way as the actual revision is signed. All values can thus be obtained by
+inspecting the identity storage.
+
+It is beyond the scope of this document to devise the exact external format to
+serialise the tuple into, as this is expected to vary from service to service.
+
+## Revocation
+
+A claim can be revoked by creating a new identity revision which simply does not
+contain the claim payload. Likewise, a later claim describing the same `SERVICE`
+invalidates an earlier one.
```

Comment:
I think temporary lack of validity is always the case -- you need to fulfill the signing obligation.

Whether or not we want to allow multiple accounts per service is an open question I'd say -- we can allow the payload type to be an array. I feel like the semantics of this is a bit wonky, though: if, for whatever reason, I choose to be one person on Radicle, but two persons on Twitter, then probably I'd like to include some indication of what's different (like "official", "shitposting" or something). But then, we'd also need to include this bit in the proof, don't we? Otherwise we'd need to try all matching.

Reply:
> I think temporary lack of validity is always the case -- you need to fulfill the signing obligation.

I think that we can solve this problem easily if we allow the opposite claims to overlap in validity time ranges only partially.

Ethereum could be a good example. Let's say that today is day 80 (of whatever). I find a claim on Ethereum ranging from day 50 to 200, looks good. I go to link and find a counterpart claim from day 50 to 100, still good, now I can consider this attestation valid until day 100. A few days pass and I get an update from link: it claims the same attestation from days 50 to 200. This is great, now I know that the attestation is valid until day 200.

This schema allows seamless updates of attestations despite lack of synchronization and guarantees that on both sides we only need to know the single latest claim to have the full knowledge.

The only downside is that the validity ranges now must be solid: they have only start, end and no gaps inside. This is probably fine, I can't think of any use case where I would need to express something like "this Twitter account will be mine until April 1st, then it'll be somebody else's for a week and after that again mine for a year".

Reply:
> I choose to be one person on Radicle, but two persons on Twitter, then probably I'd like to include some indication of what's different

If my previous comment is correct, we are guaranteed to only need 1 entry in the identity JSON per account. We could bypass the multiple accounts problem by duplicating the services, e.g. "twitter", "twitter_2" and "twitter_3". This will make the number of accounts strictly limited, but it'll be trivial to build a UI to display them and keep track of which account is which. Another advantage is that for now we can stick with having only "twitter" and add support for "twitter_2" when we fell like doing so. The JSON will stay unambiguous to interpret and easy to navigate.

Reply:
I am not sure I'm following. You **must** create a new revision always, but it may remain invalid until it was signed by the delegate quorum. There is also no guarantee that the tip will be seen by others -- that's why the revision is important: if you come across a proof, but are not able to retrieve the revision it is talking about, there is something funky going on.

It is impossible to talk about time without a reference. Arguably, when you want to have expiration, wall-clock time is _probably_ good enough -- after a week, _most_ people's computers will agree that `created + expires` is in the past. You can't safely base something like handover on this, though: if I publish such a proof on my homepage, but later sell the domain, the buyer can keep the proof up, but it would be compromised. If your time unit is block height, that problem goes away.

Reply:
> You must create a new revision always, but it may remain invalid until it was signed by the delegate quorum.

I really don't think that the quorum signing process matters here. Let's assume that all revisions not quorum-signed are invalid and thus will be ignored by everyone except the quorum members.

> the revision is important

It seems that you envision a system where a link identity revision contains a link to proof on an external service (transaction hash) and that proof (transaction content) contains a reference to said identity revision leading to a chicken and egg problem.

We've discussed that yesterday and I think that we came to a conclusion, that there are means of deciding, which link revision is the tip (from best effort perspective, there may always be a better one floating on the network). This would break the cycle and allow looser and simpler relation between claims.

> if you come across a proof, but are not able to retrieve the revision it is talking about, there is something funky going on

Does any revision except the tip matter? Even if a 3rd party claim contains a proper revision, it doesn't matter if the tip doesn't recognize the said 3rd party.

> wall-clock time is probably good enough

Absolutely. You can't verify any security without a decent clock synchronization. If it's off by a few days and the user uses it to decide if they trust something on the internet, it's their own fault.

> if I publish such a proof on my homepage, but later sell the domain, the buyer can keep the proof up, but it would be compromised

I think that there is a missing part in this example: after selling a domain I buy a new one and hand over the attestation to it, right? In that case there's nothing we can do about it, if after handing over there are still people who are looking at a claim on a compromised 3rd party and hold an old link state, they will always trust the new owner. It doesn't matter how we measure time.

> If your time unit is block height, that problem goes away.

How? BTW every block has a timestamp with precision of a few seconds, but it probably doesn't matter if the claim on Ethereum contains the validity time range submitted as a parameter.

Reply:
I don’t understand what you are debating, so I think we may just stop here.

What we’ve discussed works pretty much exactly like the ACME protocol (RFC 8555), with comparable guarantees. This is enough for the purpose, but not enough for other security aspects of a `radicle-link` identity. Therefore, this RFC stands on its own.

You are welcome to suggest improvements to this proposal, or ask to clarify aspects of it, but I would like to ask you to do so in a more systematic way — just stating that this or that doesn’t matter, or that something clearly works without explaining your axioms makes it really hard to figure out how to respond.

So, if you forget about blockchains — I have removed everything except for merkle root anchoring — do the questions you raised in your last comment persist?

Reply:
I'm trying to figure out how is the handover supposed to work? I haven't found it in the RFC, so I've proposed a solution, but you seem to have a different solution already in your mind.

Reply:
Aha, ok cool. I brought up this example trying to guess whether you are trying to find some flaw in this RFC, or considering if there needs to be anything else for the planned contract.

Maybe let’s look at the contract first: we said it’s just `register(URN)`, `unregister(URN)`, and we store the corresponding tx hashes in link. Let’s say the link revision which contains the unregister- (or simply removal of the register-, I’m not sure) tx gets lost. When I follow the register tx, I will find that it’s valid, but without an extra step I don’t know if later it was unregistered (and maybe someone else registered). However, we know that it is virtually impossible to _hide_ the chain prefix, so the window in which someone else who registered id `X` can pretend to own it is a couple of confirmations.

So I would say this is fairly safe without auto-expiration, assuming that it’s not hard to check if the _current_ address holding the link ID is the same as the one in the tx we know.


Let’s take away the contract: I post a proof somewhere where there is a risk of takeover (or just handover), like a DNS name. Let’s say I’m using LetsEncrypt. The is no way to perform the online check we just introduced, because there is no information about who owns the domain, or if that changed (that is, we cannot ask “who owned the domain at time X”). Because we could also be missing the revocation revision from link, it would be wise to specify an expiration time in the claim (preferably set to some time _before_ the domain expires).

Finally, we want to defend against account compromise (whether that’s social media or Ethereum). The only option we have is to revoke on the link side, but we cannot be sure that this is received. That’s where anchoring (or: checkpointing) the latest revision somewhere else comes in: if I see a revision there which I don’t have, I should reject any proofs until I have it. Without this, an expiration time will also do, but with larger uncertainty.

Does this make sense?